test/server/http_filter_integration_test_base.cc

```diff
@@ -0,0 +1,56 @@
+#include "test/server/http_filter_integration_test_base.h"
+
+#include "server/well_known_headers.h"
+
+#include "gtest/gtest.h"
+
+namespace Nighthawk {
+
+HttpFilterIntegrationTestBase::HttpFilterIntegrationTestBase(
+    Envoy::Network::Address::IpVersion ip_version)
+    : HttpIntegrationTest(Envoy::Http::CodecClient::Type::HTTP1, ip_version),
+      request_headers_({{":method", "GET"}, {":path", "/"}, {":authority", "host"}}) {}
+
+void HttpFilterIntegrationTestBase::setup(const std::string& config) {
+  config_helper_.addFilter(config);
+  HttpIntegrationTest::initialize();
+}
+
+void HttpFilterIntegrationTestBase::updateRequestLevelConfiguration(
+    absl::string_view request_level_config) {
+  setRequestHeader(Server::TestServer::HeaderNames::get().TestServerConfig, request_level_config);
+}
+
+void HttpFilterIntegrationTestBase::switchToPostWithEntityBody() {
+  setRequestHeader(Envoy::Http::Headers::get().Method,
+                   Envoy::Http::Headers::get().MethodValues.Post);
+}
+
+void HttpFilterIntegrationTestBase::setRequestHeader(
+    const Envoy::Http::LowerCaseString& header_name, absl::string_view header_value) {
+  request_headers_.setCopy(header_name, header_value);
+}
+
+Envoy::IntegrationStreamDecoderPtr
+HttpFilterIntegrationTestBase::getResponse(ResponseOrigin expected_origin) {
+  cleanupUpstreamAndDownstream();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  Envoy::IntegrationStreamDecoderPtr response;
+  const bool is_post = request_headers_.Method()->value().getStringView() ==
+                       Envoy::Http::Headers::get().MethodValues.Post;
+  const uint64_t request_body_size = is_post ? 1024 : 0;
+  if (expected_origin == ResponseOrigin::UPSTREAM) {
```

Comment:
I'm having trouble understanding this function's full intent.  There are a lot of conditionals that seem to make the behavior very different.  Is this actually common behavior across filters?

Maybe some comments in this section about why things are happening would be helpful.

Reply:
I added some comments in ca17f41. Does that provide the type of context you'd consider helpful when reading this code? 
Basically, this support call is an extract of the what we need to cover the stream codec code for all our extensions. We'll later see that this allows us deprecate much more complex code (which was the way to do it a while back), as wel as deduplicate some similar code.

Reply:
Additionally, for some more concrete context, https://github.com/envoyproxy/nighthawk/pull/533 is the direct follow up to this (currently in draft, as it depends on this PR, and will lean up once we merge this one and merge the result into that). It brings in a [new test file](https://github.com/envoyproxy/nighthawk/blob/97322d867ca1ebeb35bed0412e7b603bff4ba8a6/test/server/http_filter_base_test.cc) which is using this extensively. 
That test file adds tests for ensuring basic behavioural guarantees across all of our extensions: https://github.com/envoyproxy/nighthawk/blob/97322d867ca1ebeb35bed0412e7b603bff4ba8a6/test/server/http_filter_base_test.cc

The behaviours we are looking for involve certain configuration (error) handling, as well as handling default/empty configuration, both static from disk as well as dynamically delivered via response headers. These tests are parameterised over GET and POST requests.

Once we land that, we can in turn start purging / enhancing the other test files, and make them more about just testing the functionality that the extension brings in. Where applicable, it will also use the functionality we add in this PR, and we can purge their own variant(s).

Reply:
(NOTE: the changes from the last paragraph above, are already staged over at #512)